File: LeetCode/JavaScript/B17-24-Swap Nodes in Pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function (head) {
  if (!!head && !!head.next) {
    const b = head.next;
    [head.next, b.next] = [swapPairs(b.next), head];
    return b;
  }
  return head;
};

File: LeetCode/5회차/B17-24-Swap Nodes in Pairs.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        root = prev = ListNode()
        node = head
        prev.next = node

        while node and node.next:
            next, prev.next, node.next.next = node.next.next, node.next, node
            node.next = next
            prev, node = node, node.next

        return root.next


